### ISSUE TITLE
add functions for loo predictions

### ISSUE THREAD

**Issue by avehtari** [stan-dev/rstanarm#159]
#### Summary:
Add the following functions
- [x] loo_predict
- [x] loo_linpred
- [x] loo_predictive_interval

#### Description:
loo_predict, loo_linpred, and loo_predictive_interval would correspond to posterior_predict, posterior_linpred and predictive_interval functions, but would produce results using loo predictive distributions. These would be useful for making additional loo predictive performance estimates with other than log score utility/cost functions and for LOO-PIT predictive model checking.

Example code

```
data(radon)
y<-radon$log_radon
# Fit the first model
modelA <- stan_lmer(
    log_radon ~ floor + log_uranium + floor:log_uranium + (1 + floor | county),
    data = radon,
    cores = 4,
    iter = 2000,
    chains = 4)
    
# loo_linpred in 4 lines
log_likA<-log_lik(modelA, parameter_name = "log_lik")
psisA<-psislw(-log_likA)
linpredsA<-posterior_linpred(modelA)
loolinpreds<-colSums(linpredsA*exp(psisA$lw_smooth))

# loo_predict in 4 lines
log_likA<-log_lik(modelA, parameter_name = "log_lik")
psisA<-psislw(-log_likA)
predsA<-posterior_predict(modelA)
loopredsA=colSums(predmsA*exp(psisA$lw_smooth))

# loo predictive interval can be computed from the above results

# probability integral transformation (PIT)
# this would be more accurate using conditional cdf's N(y_i|mu^{(s)},sigma^{(s)}
pitA<-array(0,ncol(predsA))
library(matrixStats)
for (i in 1:ncol(predsA)) {
    pitA[i]<- exp(logSumExp(psisA$lw_smooth[(predsA[,i]<=y[i]),i]))
}

# LOO-PITs should have uniform distribution
par(mfrow=c(1,2))
qqplot(pitA,runif(10000))

```

**Comment by jgabry** [comment]
Great, thanks for opening the issue Aki.

**Comment by jgabry** [comment]
@avehtari Would the prefix `psis_` make more sense than the prefix `loo_` for these new function names? I like `loo_` better but `psis_` seems more descriptive. What do you think? 

Another option would be to not make new functions and instead add an argument to the existing functions (`posterior_predict`, `poster_linpred`, `predictive_interval`) that turns on PSIS-weighting. @bgoodri Do you have a preference?

**Comment by avehtari** [comment]
I think loo is better, as we might end up running also PSIS-LOO+, that is, computing exact LOO for those observations which have k>0.7. I'm also find with an additional argument as it probably would reduce duplication of documentation and code.

**Comment by jgabry** [comment]
Ok cool. We'll either add functions with the "loo" prefix or add a T/F
argument "loo" that toggles whether or not to weight the output before
returning it.

**Comment by jgabry** [comment]
@avehtari Another question about these functions. For `loo_linpred`, for example, your code above returns a vector (the column sums of the psis-weighted linear predictor matrix), but the `posterior_linpred` function returns a matrix. Is there a reason why it shouldn't return the full weighted matrix? 

If you want the loo versions to only a point estimate (vector) then I think we should make separate functions because the return type of `loo_linpred` and `posterior_linpred` will be different, but if it returns the weighted matrix then I think we should just add a `loo` argument to the existing `posterior_linpred` function to toggle whether or not that matrix is weighted before returning it. Same thing applies to `posterior_predict`.

**Comment by paul-buerkner** [comment]
@avehtari Do you have a reference for what these functions can be used later on? I ask because I am thinking about implementing them in brms as well (as soon as you have agreed on the names so we can use the same names across packages)

**Comment by jgabry** [comment]
@paul-buerkner Well, one thing for sure is these plots that I will add to __bayesplot__:  https://github.com/stan-dev/bayesplot/issues/72

**Comment by avehtari** [comment]
@jgabry Ah, posterior_linpred returns posterior samples, but the above code computes the psis-weighted expectation. Full weighted matrix would be just separately the matrix of posterior samples and matrix of weights. It is possible to use resampling to get loo posterior samples (sample with replacement), and it might be sometimes useful for someone but not often. Resampling causes loosing some information, and it's better to use weighting to compute any expectations we would like to use (mean, variance, quantiles, etc.). So it seems it's then better to have a separate loo-functions which would get also the function (mean, variance, quantiles, etc) as an argument which expectation we want to compute (with mean as default).

@paul-buerkner stan-dev/bayesplot#72 as mentioned by @jgabry (while I was writing this) and PIT mentioned above can be used for predictive checking

**Comment by jgabry** [comment]
Ok thanks, that makes sense Aki. I think that's the best way to go.

Paul, I will also add generics to rstantools for these so that both
rstanarm and brms can use those.

**Comment by paul-buerkner** [comment]
@jgabry @avehtari thanks for the reference and for adding generics to rstantools.

**Comment by jgabry** [comment]
@avehtari You say above: 

> So it seems it's then better to have a separate loo-functions which would get also the function (mean, variance, quantiles, etc) as an argument which expectation we want to compute (with mean as default).

Do you mean that the argument would let the user pass in an actual R function or that we would support a few common ones and only allow selecting from that list? 

For example, we could have the following signature for `loo_predict` (and `loo_linpred`)

```r
#' @param object Fitted model object.
#' @param lw Matrix of smoothed log-weights. 
#' @param fun A function computing the desired expectation. 
loo_predict(object, lw, fun) 
```
where `fun` is an actual function (e.g. `sum` in your example) specified by the user. 
Or we could have or something like this 

```r
#' @param type Which expectation to compute. Can be "mean", "variance", "quantile" (I guess with extra arg 'prob').
loo_predict(object, lw, type = "mean") 
```
where `type` only lets the user select from a few options.

**Comment by avehtari** [comment]
I think it's not easy allow any user specified function. For example, to compute variance using importance sampling weight vector w and data vector x:
IS-weighted variance: Var_IS(x)=sum(w*(x^2))-sum(w*x)^2
And for quantile you need to solve the inverse cdf using the weighted empirical CDF

**Comment by jgabry** [comment]
Ok great, glad that's what you were thinking because I don't know how we
would do it the first way! Allowing just mean, var/sd, and quantiles as the
expectations the user can select will be much easier.

On Thu, Mar 16, 2017 at 2:56 PM Aki Vehtari <notifications@github.com>
wrote:

I think it's not easy allow any user specified function. For example, to
compute variance using importance sampling weight vector w and data vector
x:
IS-weighted variance: Var_IS(x)=sum(w*(x^2))-sum(w*x)^2
And for quantile you need to solve the inverse cdf using the weighted
empirical CDF

—
You are receiving this because you were mentioned.


Reply to this email directly, view it on GitHub
<https://github.com/stan-dev/rstanarm/issues/159#issuecomment-287157591>,
or mute the thread

**Comment by bgoodri** [comment]
Can this be closed now that the PR is merged?

**Comment by jgabry** [comment]
I think it can be closed but we should open a separate issue for extending
this functionality to PSIS-LOO+ and Kfold, which will require more work.
Right now it only works for regular PSIS.

On Tue, Apr 11, 2017 at 4:55 AM bgoodri <notifications@github.com> wrote:

Can this be closed now that the PR is merged?

—
You are receiving this because you were mentioned.


Reply to this email directly, view it on GitHub
<https://github.com/stan-dev/rstanarm/issues/159#issuecomment-293194423>,
or mute the thread

**Comment by jgabry** [comment]
closed by PR #176
